Sanitize user `full_name` without roadblocks

bickelj opened the pull request:
Rather than impeding the flow of users setting full name or requiring a new variable to be set by administrators, clean up users' given full names. This means removing all HTML tags and removing colons and slashes. It may be unfriendly to people such as "David-Wynn: Miller" but this is the cost of nerfing spam URLs in full names.

Issue #4060

## Test Steps

 - [ ] Create a user with the full name `<a href="https://myspam.com">Somebody: somewhere</a>`
 - [ ] The resulting user name should be `Somebody somewhere` (no HTML tags, no colons, no slashes).

frankduncan committed:
Add validation check and configuration for user full names

Set `INVALID_FULL_NAME_REGEXES` to a list containing regular
expressions that when matched should cause `full_name` to be rejected.

Issue #4060

frjo commented:
@bickelj Can we not run username through `nh3.clean()` instead?

bickelj commented:
@frjo Perhaps, but from what I see at https://nh3.readthedocs.io/en/stable/ and https://github.com/rust-ammonia/ammonia the library appears to be for "HTML sanitization" where the output string is expected to be HTML. In this case, I think we are looking for plain text both on input and on output. `nh3` might be useful additionally but the goal here is to disallow URLs completely.

Edit: I see `nh3.is_html` which we could use to reject any HTML tags, is that what you had in mind? But that would still allow non-HTML URLs to come through, correct?

bickelj committed:
Prohibit HTML in `full_name` field

Prior to looking for prohibited regular expression patterns, check for
HTML tags in the `full_name` and invalidate when HTML is found.

Issue #4060

bickelj commented:
ffb57c89e09d10c4e7d7871b6e6eed3f2bbf4d34 uses `nh3` to prohibit HTML content in a Full name.

frjo commented:
@bickelj It can clean out all html.

```
>>> import nh3
>>> nh3.clean("<b><a href='https://example.com'>Hello</a></b>", tags=set())
'Hello'
```

But that leaves all url schemas intact, the original problem needed to be solved.

Could we add a hardcoded feature that removes every ":" and "/"? That would break all links and they should not be in any real usernames in any case.

I'm open to doing this silently. If the user only adds tags etc. they end up with a empty username. No error message or anything. Will only effect people with bad intensions.

bickelj commented:
@frjo Ah, right. That makes sense. Let me try that.

bickelj committed:
Sanitize user `full_name` without roadblocks

Rather than impeding the flow of users setting full name or requiring
a new variable to be set by administrators, clean up users' given full
names. This means removing all HTML tags and removing colons and
slashes. It may be unfriendly to people such as "David-Wynn: Miller"
but this is the cost of nerfing spam URLs in full names.

Issue #4060

bickelj commented:
@frjo Updated. It looks like it's in two places, now, though. I guess an API should be able to create users as well which might omit the form but would touch the model. Perhaps the sanitation should only be in the model. What do you think?

Edit: and obviously needs fixed tests now.

bickelj commented:
@frjo I found this test that clearly intends to allow weird names including HTML and such:
```python
    def test_doesnt_mess_with_name(self):
        full_name = "I have; <a> wei'rd name"
        self.submit_form(name=full_name)
        submission = ApplicationSubmission.objects.first()
        self.assertEqual(submission.user.full_name, full_name)
```

I suppose we should change the test to indicate the opposite policy.

bickelj commented:
The plot thickens: that test passes. Ah. That is on the application submission, not on the user profile, perhaps.

frjo commented on hypha/apply/users/forms.py:
```diff
@@ -198,6 +199,15 @@ def clean_email(self):
+        # Remove all colons and slashes to nerf URLs regardless of HTML tags.
+        cleaned_full_name = cleaned_full_name.replace(":", "")
+        cleaned_full_name = cleaned_full_name.replace("/", "")
+        return cleaned_full_name
```
Looks good but can we break out this into a util function? Then we have this logic only in one place and can reuse it where needed.

bickelj replied:
Good call. I left the function in the same file because this is its one use. If it needs to be moved elsewhere in future due to use outside this file or speculatively now that is also fine.

frjo commented on hypha/settings/base.py:
```diff
@@ -155,6 +155,7 @@
 # Require an applicant to view their rendered application before submitting
 SUBMISSION_PREVIEW_REQUIRED = env.bool("SUBMISSION_PREVIEW_REQUIRED", True)
 
+
```
Can we clear out this please.

bickelj replied:
I did it in a later commit because I assume these get squashed. If I need to amend, let me know.

frjo commented on hypha/apply/users/forms.py:
```diff
@@ -156,6 +157,15 @@ def __init__(self, register_view=False, request=None, *args, **kwargs):
+    # Remove all colons and slashes to nerf URLs regardless of HTML tags.
+    cleaned_value = cleaned_value.replace(":", "")
+    cleaned_value = cleaned_value.replace("/", "")
+    return cleaned_value
```
Can we move this to `hypha/apply/users/utils.py`?

frjo commented on hypha/apply/users/models.py:
```diff
@@ -170,6 +171,13 @@ def get_or_create_and_notify(
+                # Remove all colons and slashes to nerf URLs regardless of HTML tags.
+                cleaned_full_name = cleaned_full_name.replace(":", "")
+                cleaned_full_name = cleaned_full_name.replace("/", "")
+                kwargs["full_name"] = cleaned_full_name
```
Please use `strip_html_and_nerf_urls()` here as well.

bickelj replied:
Doh! Done...

bickelj committed:
Extract a field cleaning function

It may be generally useful to strip HTML and nerf URLs, though
removing them from `full_name` is the first use.

Issue #4060

bickelj commented on hypha/apply/users/utils.py:
```diff
@@ -1,5 +1,6 @@
 import string
 
+import nh3
```
My IDE complains that we do not have `nh3` in the project requirements. We have `django-nh3`. Are we importing a transitive dep here?

Yes:
```
$ grep nh3 requirements.txt 
django-nh3==0.1.1
$ pip show nh3
Name: nh3
Version: 0.2.18
...
Required-by: django-nh3
```

Maybe if we import it we should add it.

bickelj replied:
Added in 3ef7a738daee00c0555c280734a09a1139de8295

bickelj committed:
Declare `nh3` direct dependency

Hypha uses `nh3` in imports in several places. Previous to this commit
`nh3` was a transitive dependency on `django-nh3`. Instead of using an
implicitly present dependency, mark it directly. This is clearer and
also allows us to set a version where `django-nh3` did not.

Issue #4060

wes-otf commented:
works well, thanks @bickelj!
